lua/nvim-treesitter/refactor/navigation.lua

```diff
@@ -25,32 +25,85 @@ end
 
 function M.goto_definition_lsp_fallback(bufnr) M.goto_definition(bufnr, vim.lsp.buf.definition) end
 
+-- Get definitions (unique and sorted).
+local function get_definitions(bufnr)
+  local definitions = locals.get_definitions(bufnr)
+  local nodes_set = {}
+
+  for _, def in ipairs(definitions) do
+    locals.recurse_local_nodes(def, function(_, node, _, match)
+      -- lua doesn't support lookup of tables by value,
+      -- use the value from byte count instead.
+      local _, _, start = node:start()
+      nodes_set[start] = {node = node, type = match or ""}
+    end)
+  end
+
+  local definition_nodes = {}
+  for _, node in pairs(nodes_set) do
+    table.insert(definition_nodes, node)
+  end
+
+  table.sort(definition_nodes, function (a, b)
+    local _, _, start_a = a.node:start()
+    local _, _, start_b = b.node:start()
+    return start_a < start_b
+  end)
+
+  return definition_nodes
+end
+
 function M.list_definitions(bufnr)
   local bufnr = bufnr or api.nvim_get_current_buf()
-  local definitions = locals.get_definitions(bufnr)
+  local definitions = get_definitions(bufnr)
 
   if #definitions < 1 then return end
 
   local qf_list = {}
 
-  for _, def in ipairs(definitions) do
-    locals.recurse_local_nodes(def, function(_, node, _, match)
-      local lnum, col, _ = node:start()
-
-      table.insert(qf_list, {
-        bufnr = bufnr,
-        lnum = lnum + 1,
-        col = col + 1,
-        text = ts_utils.get_node_text(node)[1] or "",
-        kind = match and match:sub(1, 1) or ""
-      })
-    end)
+  for _, node in ipairs(definitions) do
+    local lnum, col, _ = node.node:start()
+    local type = string.upper(node.type:sub(1, 1))
+    table.insert(qf_list, {
+      bufnr = bufnr,
+      lnum = lnum + 1,
+      col = col + 1,
+      text = ts_utils.get_node_text(node.node)[1] or "",
+      type = type,
+    })
   end
 
   vim.fn.setqflist(qf_list, 'r')
   api.nvim_command('copen')
 end
 
+function M.list_definitions_toc(bufnr)
+  local winnr = api.nvim_get_current_win()
+  local bufnr = api.nvim_win_get_buf(winnr)
+  local definitions = get_definitions(bufnr)
+
+  if #definitions < 1 then return end
+
+  local loc_list = {}
+
+  for _, node in ipairs(definitions) do
```

Comment:
This seems to duplicate some logic in the above function. Can we abstract this to be more DRY?

Reply:
The code changed, not sure if is worth to reuse now